File: Contest/second_year_contests/contest_3/E_Valeriy_and_Deque.py
n, q = map(int, input().split())
nums = list(map(int, input().split()))
queries = {}

maxx = 0
for _ in range(q):
    query = int(input())
    queries[query] = None
    maxx = max(maxx, query)

for i in range(maxx):
    if i + 1 in queries:
        queries[i + 1] = [nums[0], nums[1]]
    A = nums.pop(0)
    B = nums.pop(0)

    if A > B:
        nums.append(B)
        nums.insert(0, A)
    else:
        nums.append(A)
        nums.insert(0, B)

for pair in queries.values():
    print(*pair)

File: Contest/second_year_contests/contest_3/E_Valeriy_and_Deque.cpp
#include <iostream>
#include <vector>
#include <map>
#include <algorithm>

int main() {
    int n, q;
    std::cin >> n >> q;

    std::vector<int> nums(n);
    for (int i = 0; i < n; i++) {
        std::cin >> nums[i];
    }

    std::map<int, std::vector<int>> queries;

    int maxx = 0;
    for (int i = 0; i < q; i++) {
        int query;
        std::cin >> query;
        queries[query] = std::vector<int>();
        maxx = std::max(maxx, query);
    }

    for (int i = 0; i < maxx; i++) {
        if (queries.find(i + 1) != queries.end()) {
            queries[i + 1] = {nums[0], nums[1]};
        }
        int A = nums[0];
        int B = nums[1];

        nums.erase(nums.begin(), nums.begin() + 2);

        if (A > B) {
            nums.push_back(B);
            nums.insert(nums.begin(), A);
        } else {
            nums.push_back(A);
            nums.insert(nums.begin(), B);
        }
    }

    for (const auto& pair : queries) {
        for (int num : pair.second) {
            std::cout << num << " ";
        }
        std::cout << std::endl;
    }

    return 0;
}